### ISSUE TITLE
parallelity of queries (multiple queries at once)?

### ISSUE THREAD

**Issue by andredoering** [brianc/node-postgres#1948]
I wrote a program, which requests n-queries (for instance 20,000) at once to my postgres database.

When I do this with **several clients** who want to query 20,000 at once too, there is **no parallelity**. That means, the requests of the second client will be queued until the first client finished all his queries.

Is this a normal behavior for postgres? If yes, how can I prevent that one user gets all the ressources (and the others have to wait) if there is no **parallelity**?

This is my code:
```
const { Pool } = require("pg");
const pool = new Pool();

pool.connect((err, client, done) => {
  if (err) throw err;
  client.query("SELECT * from member where m_id = $1", [1], (err, res) => {
    done();
    if (err) {
      console.log(err.stack);
    } else {
      console.log(res.rows[0]);
    }
  });
});  
```

**Comment by charmander** [comment]
When you say “several clients”, does that mean several requests to a single server’s `/`? If so, that’s expected, yes – all 20,000 `benchmark()`s attempt to take clients from the pool at the same time. The first 10 (the default maximum pool size) create connections and the other 19,990 wait in the queue. If another 20,000 `benchmark()`s come in during that time, they enter at the end of the queue.

There are lots of different strategies you could use to change this behaviour. Which ones are good choices depends on your specific situation. For example, you could give the next query to each waiting client in turn by advancing through a set of queues representing the active requests for `/`.

Of course, you should also make sure you need to run 20,000 queries at all. That wouldn’t be a good way to select 20,000 rows by id.

**Comment by andredoering** [comment]
Thanks for your answer. And yes several clients means for instance several web clients who access the server route '/'. 

I am not exactly sure what you mean by saying:

> For example, you could give the next query to each waiting client in turn by advancing through a set of queues representing the active requests for /.

Can you explain it to me more (maybe with an example)?

I am running this benchmark, because I want to know how my node server will behave, when I go in production later and have multi web clients accessing a route and querying the database. How is it usually handled?

**Comment by charmander** [comment]
> I am running this benchmark, because I want to know how my node server will behave, when I go in production later and have multi web clients accessing a route and querying the database. How is it usually handled?

Does a route involve querying the database 20,000 times? Typically (a small number of queries per request) the default queuing behaviour is fine.

> Can you explain it to me more (maybe with an example)?

Maybe later. It’s a bit involved.

**Comment by andredoering** [comment]
Okay Thank you. 

> Does a route involve querying the database 20,000 times? Typically (a small number of queries per request) the default queuing behaviour is fine.

No it doesn't. I wanted to write an extreme benchmark example. I wanted to monitor what happens, if I have a large amount of users who request routes with database queries.

What happens if a user wants to abuses my endpoints and requests routes thousands of times a second? That would freeze my database, since all the others have to wait, right?

Wouldn't it be an option to set the pool size to 1 Million? lol

**Comment by charmander** [comment]
> What happens if a user wants to abuses my endpoints and requests routes thousands of times a second? That would freeze my database, since all the others have to wait, right?

Yes, but this isn’t a problem that a different way of scheduling work can solve for you. Avoiding DoS is one of the hardest things to do. Start by looking into rate limiting.

Also, a more realistic test would be to make, say, 4,000–20,000 times as many requests with 1–5 queries per request.

> Wouldn't it be an option to set the pool size to 1 Million? lol

You can set it higher, but at a certain point (much lower than one million) it doesn’t help. The database server only has so much capacity itself.

**Comment by andredoering** [comment]
> You can set it higher, but at a certain point (much lower than one million) it doesn’t help. The database server only has so much capacity itself.

I tried to change the pool Size, but it doesn't work, what do I wrong? No matter what I set up as a poolSize, it has the same behavior and when I access to clients who are waiting with     `console.log(pool.waitingCount);` I get the same results.

This is my code:

```
const express = require('express');
const app = express();

const pg = require("pg");
pg.defaults.poolSize = 20;
const pool = new pg.Pool();

benchmark(){
	pool.connect((err, client, done) => {
	  if (err) throw err;
	  client.query("SELECT * from member where m_id = $1", [1], (err, res) => {
	    done();
	    if (err) {
	      console.log(err.stack);
	    } else {
	      console.log(res.rows[0]);
	    }
	  });
	});	
}


app.get('/', function(req, res) {
  for(let i=0;i<20000;i++){
	benchmark();
  }
});
```

Thank you very much for your answers so far!

**Comment by charmander** [comment]
`Pool` doesn’t use `pg.defaults`. Pass options:

```js
const pool = new pg.Pool({ max: 20 });
```

[documentation](https://node-postgres.com/api/pool/#constructor)

**Comment by andredoering** [comment]
Thank you for your replies you were very helpful!
